lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -244,28 +243,54 @@ static Value *mergeDwordsIntoVector(IRBuilder<> &builder, ArrayRef<Value *> inpu
   return vec;
 }
 
+// =====================================================================================================================
+// Lower as.continuation.reference call.
+//
+// @param asCpsReferenceOp: the instruction
+void PatchEntryPointMutate::lowerAsCpsReference(cps::AsContinuationReferenceOp &asCpsReferenceOp) {
+  IRBuilder<> builder(&asCpsReferenceOp);
+  auto *ref = builder.CreatePtrToInt(asCpsReferenceOp.getFn(), builder.getInt32Ty());
+
+  Function &callee = cast<Function>(*asCpsReferenceOp.getFn());
+  auto level = cps::getCpsLevelFromFunction(callee);
+  ref = builder.CreateOr(ref, (uint64_t)level);
+
+  asCpsReferenceOp.replaceAllUsesWith(ref);
+  asCpsReferenceOp.eraseFromParent();
+}
+
 // =====================================================================================================================
 // Lower calls to cps function as well as return instructions.
 //
 // @param func : the function to be processed
 bool PatchEntryPointMutate::lowerCpsOps(Function *func) {
-  SmallVector<CallInst *> cpsOps;
+  SmallVector<cps::JumpOp *> cpsJumps;
+  SmallVector<cps::AsContinuationReferenceOp *> asCpsRefs;
   struct Payload {
-    SmallVectorImpl<CallInst *> &calls;
+    SmallVectorImpl<cps::JumpOp *> &jumps;
+    SmallVectorImpl<cps::AsContinuationReferenceOp *> &cpsRefs;
   };
-  Payload payload = {cpsOps};
-  static auto addCpsCall = [](Payload &payload, CallInst *op) { payload.calls.push_back(op); };
-
-  static auto visitor = llvm_dialects::VisitorBuilder<Payload>()
-                            .setStrategy(llvm_dialects::VisitorStrategy::ByFunctionDeclaration)
-                            .add<cps::JumpOp>([](auto &payload, auto &op) { addCpsCall(payload, &op); })
-                            .build();
+  Payload payload = {cpsJumps, asCpsRefs};
+
+  static auto visitor =
+      llvm_dialects::VisitorBuilder<Payload>()
+          .setStrategy(llvm_dialects::VisitorStrategy::ByFunctionDeclaration)
+          .add<cps::JumpOp>([](auto &payload, auto &op) { payload.jumps.push_back(&op); })
+          .add<cps::AsContinuationReferenceOp>([](auto &payload, auto &op) { payload.cpsRefs.push_back(&op); })
```

Comment:
Do we need to collect the AsContinuationReference ops or can we immediately lower them when found?

Reply:
That's a good suggestion! It helps simplify the code and avoid the long comment. I have incorporated your suggestion in the new version.